clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -293,6 +276,23 @@ private void process(final ConsumerRebalanceListenerCallbackNeededEvent event) {
     private final AtomicLong currentThread = new AtomicLong(NO_CURRENT_THREAD);
     private final AtomicInteger refCount = new AtomicInteger(0);
 
+    private final InvalidGroupIdException invalidGroupIdException =
+        new InvalidGroupIdException("To use the group management or offset commit APIs, you must " +
+            "provide a valid " + ConsumerConfig.GROUP_ID_CONFIG + " in the consumer configuration.");
+
+    private void updateGroupMetadata(final Optional<Integer> memberEpoch, final Optional<String> memberId) {
+        groupMetadata.updateAndGet(
```

Comment:
Can we not just use a `volatile` variable here? We only have the background thread updating the epoch and the ID, with the exception of the shutdown procedure in the application thread. A race between background thread and application thread however doesn't really seem to be solved by using `updateAndGet` (that would require checking if the consumer group metadata is "closed" here).

Reply:
The application thread can also update the group metadata when unsubscribing.